ducktape/mark/_mark.py

```diff
@@ -0,0 +1,374 @@
+# Copyright 2015 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from ducktape.tests.test import TestContext
+from ducktape.errors import DucktapeError
+
+import functools
+import itertools
+
+
+class Mark(object):
+    """Common base class for "marks" which may be applied to test functions/methods."""
+
+    @staticmethod
+    def mark(fun, mark):
+        """Attach a tag indicating that fun has been marked with the given mark
+
+        Marking fun updates it with two attributes:
+
+        - marks:      a list of mark objects applied to the function. These may be strings or objects subclassing Mark
+                      we use a list because in some cases, it is useful to preserve ordering.
+        - mark_names: a set of names of marks applied to the function
+        """
+        # Update fun.marks
+        if hasattr(fun, "marks"):
+            # This ordering makes the ordering of marks match
+            # the way parametrize annotations are stacked
+            fun.marks.insert(0, mark)
+        else:
+            fun.__dict__["marks"] = [mark]
+
+        # Update fun.mark_names
+        if hasattr(fun, "mark_names"):
+            fun.mark_names.add(mark.name)
+        else:
+            fun.__dict__["mark_names"] = {mark.name}
+
+    @staticmethod
+    def marked(f, mark):
+        if f is None:
+            return False
+
+        if not hasattr(f, "mark_names"):
+            return False
+
+        return mark.name in f.mark_names
+
+    @staticmethod
+    def clear_marks(f):
+        if not hasattr(f, "marks"):
+            return
+
+        del f.__dict__["marks"]
+        del f.__dict__["mark_names"]
+
+    @property
+    def name(self):
+        return "MARK"
+
+    def apply(self, seed_context, context_list, maps):
+        for m in maps:
+            context_list = map(m, context_list)
+        return seed_context, context_list, maps
+
+    def __eq__(self, other):
+        if type(self) != type(other):
+            return False
+
+        return self.name == other.name
+
+
+class Ignore(Mark):
+    """Ignore a specific parametrization of test."""
+    def __init__(self, **kwargs):
+        # Ignore tests with injected_args matching self.injected_args
+        self.injected_args = kwargs
+
+    @property
+    def name(self):
+        return "IGNORE"
+
+    def apply(self, seed_context, context_list, maps):
```

Comment:
Why jump through all these hoops with this `apply` method with `maps` parameter? Why wouldn't something like

```
def apply(self, contexts):
    return ctx.clone(ignore=(self.injected_args is None or self.injected_args == ctx.injected_args)) for ctx in contexts
```

work? I was thinking that every `Mark` would take a list/generator of `TestContext`s and output an updated lists. `@ignore` returns the same number, just tagged as necessary; `@parametrize` just needs to add one new entry to the end; `@matrix` does the same, etc.

If I'm reading this correctly, you've done something similar, but where you collect all the mapping functions first, then apply them all in the base class's `apply`, which makes it a lot harder to understand what's going on.